tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,711 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+// Segment header fields constants.
+const (
+	// MagicHeadChunks is 4 bytes at the head of a segment file.
```

Comment:
segment file? How we name things here? Let's decide and be consistent (: 

I thought segment was `chunk/000001` file? Can we define clear naming?

Reply:
I have seen "segment file" being used both for chunk files and the WAL files. So it felt natural to use the same here.

Reply:
Hm. I think that's creating just confusion. Let's change it then? (:

Reply:
Open for suggestions :)

Reply:
I think we need to start some quick discussion on -developers for that.

> and the WAL files. 

Hm... so there is 3rd segment?

I would propose:

* `index`: **Index File**
* `chunks/000001`: **Chunks File / Segment **
* `wal/chunks/000001`: **Head Chunks File / Head Segment**

But not sure about this 3rd segment you mentioned, what exactly you mean?

Reply:
> But not sure about this 3rd segment you mentioned, what exactly you mean?

https://github.com/prometheus/prometheus/blob/46129fdebbbc1f8c73b450b3629b92bfad6b2d95/tsdb/wal/wal.go#L75-L80
WAL in itself consists of multiple files where each file is called a segment :P

Reply:
:cry:

Reply:
ok so

I would propose:

* `index`: **Index File**
* `chunks/000001`: **Chunks File / Segment**
* `wal/chunks/000001`: **Head Chunks File / WAL Chunks Segment**
* https://github.com/prometheus/prometheus/blob/46129fdebbbc1f8c73b450b3629b92bfad6b2d95/tsdb/wal/wal.go#L75-L80: **WAL Segment**

Reply:
I have no strong opinions on that. I will use "head chunk file" then and replace wherever I find the word "segment"